### ISSUE TITLE
Loss won't converge using the LION optimizers

### ISSUE THREAD

**Issue by hankcs** [mosaicml/llm-foundry#317]
## Environment

```
Collecting system information...
---------------------------------
System Environment Report        
Created: 2023-06-12 23:25:33 CST
---------------------------------

PyTorch information
-------------------
PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: glibc-2.17

Python version: 3.10.11 (main, Apr 20 2023, 19:02:41) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.83.1.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-PCIE-40GB
GPU 1: NVIDIA A100-PCIE-40GB
GPU 2: NVIDIA A100-PCIE-40GB
GPU 3: NVIDIA A100-PCIE-40GB
GPU 4: NVIDIA A100-PCIE-40GB
GPU 5: NVIDIA A100-PCIE-40GB
GPU 6: NVIDIA A100-PCIE-40GB
GPU 7: NVIDIA A100-PCIE-40GB

Nvidia driver version: 515.65.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] pytorch-ranger==0.1.1
[pip3] torch==1.13.1
[pip3] torch-optimizer==0.3.0
[pip3] torchaudio==0.13.1
[pip3] torchmetrics==0.11.3
[pip3] torchtext==0.14.1
[pip3] torchvision==0.14.1
[conda] blas                      1.0                         mkl    defaults
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2023.1.0         h6d00ec8_46342    defaults
[conda] mkl-service               2.4.0           py310h5eee18b_1    defaults
[conda] mkl_fft                   1.3.6           py310h1128e8f_1    defaults
[conda] mkl_random                1.2.2           py310h1128e8f_1    defaults
[conda] numpy                     1.24.3          py310h5f9d8c6_1    defaults
[conda] numpy-base                1.24.3          py310hb5e798b_1    defaults
[conda] pytorch                   1.13.1          py3.10_cuda11.7_cudnn8.5.0_0    pytorch
[conda] pytorch-cuda              11.7                 h778d358_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] pytorch-ranger            0.1.1                    pypi_0    pypi
[conda] torch-optimizer           0.3.0                    pypi_0    pypi
[conda] torchaudio                0.13.1              py310_cu117    pytorch
[conda] torchmetrics              0.11.3                   pypi_0    pypi
[conda] torchtext                 0.14.1                   pypi_0    pypi
[conda] torchvision               0.14.1              py310_cu117    pytorch


Composer information
--------------------
Composer version: 0.14.1
Composer commit hash: None
Host processor model name: AMD EPYC 7352 24-Core Processor
Host processor core count: 48
Number of nodes: 1
Accelerator model name: NVIDIA A100-PCIE-40GB
Accelerators per node: 1
CUDA Device Count: 8
```

## To reproduce

I'm training an 1B MPT model and the Adam optimizer worked quite well but the LIONs didn't.

LION config:

```
optimizer:
  name: decoupled_lionw
  lr: 1e-4
  betas:
    - 0.9
    - 0.99
  weight_decay: 0.0
  outlier_threshold: 5
```

which is suggested by the LION paper except for the `weight_decay`. The paper was using 1e-2 which was complained by composer to be too large and it didn't yield better loss in early runs.

On a custom corpus with customized tokenizer, the `decoupled_adamw` worked quite well but the LION optimizers failed to converge anywhere close to Adam.

<img width="607" alt="hankcs com 2023-06-12 at 10 33 10 AM" src="https://github.com/mosaicml/llm-foundry/assets/5326890/dd2ed56c-1fad-45d6-b52b-740b11bb90f5">

## Expected behavior

Coverge to a loss close to Adam.

## Additional context

Is there a bug in LION or do you have any suggestions regarding hyper-parameter tuning with LION?
Thanks.

**Comment by dblalock** [comment]
Hi @hankcs,

It's hard to diagnose convergence issues with certainty, but here are a couple likely causes:

1. The lack of weight decay will cause the weights to only ever grow, rather than reach an equilibrium norm. See, e.g., [here](https://dblalock.substack.com/i/126042286/rotational-optimizers-simple-and-robust-dnn-training) for an explanation. You might not see a difference early in training, but I would expect this to cause issues eventually (as you're observing).
2. Your other hyperparameters might be causing issues. In particular, momentum [effectively scales the updates by 1/(1 - β2)](https://myrtle.ai/learn/how-to-train-your-resnet-5-hyperparameters/), which is 100 for β2 = .99. For our 1B models, we tend to use β2 = .95, which only scales by 20. Here are some hparams we've used for our MPT 1B runs:

```yaml
  optimizer:
    name: decoupled_lionw
    lr: 0.0002
    betas:
    - 0.9
    - 0.95
    weight_decay: 0.0002
```

**Comment by germanjke** [comment]
```
optimizer:
  name: decoupled_lionw
  lr: 2.0e-4
  betas:
  - 0.9
  - 0.95
  weight_decay: 0.0
```
I'm using this and my model mpt1b converges good for ~1 day with 2048 seq len & batch size 1024
but after loss spikes as well

<img width="1153" alt="image" src="https://github.com/mosaicml/llm-foundry/assets/43114292/1592ec8a-093e-433f-9a94-9b032f077500">

looks like your models have some problems with converge? 
llm foundry with opt/gpt looks good

**Comment by germanjke** [comment]
my adamw explodes as well 
I see @hankcs your adamw explodes too here

**Comment by hankcs** [comment]
> my adamw explodes as well I see @hankcs your adamw explodes too here, why you said it's works quite well ? :D or you using technique to continue training with checkpoints before explode?

Hi @germanjke , I mean it worked quite well in comparison to LION. I heard that the loss spike is an inevitable problem in LLM pretraining so it's not something we can solve immediately. Usually, we have to skip a few batches to bypass the spike.

**Comment by hankcs** [comment]
> Hi @hankcs,
> 
> It's hard to diagnose convergence issues with certainty, but here are a couple likely causes:
> 
> 1. The lack of weight decay will cause the weights to only ever grow, rather than reach an equilibrium norm. See, e.g., [here](https://dblalock.substack.com/i/126042286/rotational-optimizers-simple-and-robust-dnn-training) for an explanation. You might not see a difference early in training, but I would expect this to cause issues eventually (as you're observing).
> 2. Your other hyperparameters might be causing issues. In particular, momentum [effectively scales the updates by 1/(1 - β2)](https://myrtle.ai/learn/how-to-train-your-resnet-5-hyperparameters/), which is 100 for β2 = .99. For our 1B models, we tend to use β2 = .95, which only scales by 20. Here are some hparams we've used for our MPT 1B runs:
> 
> ```yaml
>   optimizer:
>     name: decoupled_lionw
>     lr: 0.0002
>     betas:
>     - 0.9
>     - 0.95
>     weight_decay: 0.0002
> ```

Hi @dblalock , thank you for your prompt reply. I'll try your tips!

**Comment by germanjke** [comment]
hi @hankcs 
which Adam params are you using?
have you changed `eps`?
as I know, if we put `eps` something like 0.1 0.01 0.001 we can bypass spikes

**Comment by hankcs** [comment]
> hi @hankcs which Adam params are you using? have you changed `eps`? as I know, if we put `eps` something like 0.1 0.01 0.001 we can bypass spikes

I'm using a very small `eps`, the default one in llm-foundy yaml files.

```
optimizer:
  name: decoupled_adamw
  lr: 2.0e-4
  betas:
  - 0.9
  - 0.95
  eps: 1.0e-08
  weight_decay: 0.0
```

Good to know a larger `eps` could help.

**Comment by germanjke** [comment]
@hankcs I guess little spikes are ok in pre-training, but if after explode loss going to high lever it's shows something wrong to your training progress?

**Comment by germanjke** [comment]
> > hi @hankcs which Adam params are you using? have you changed `eps`? as I know, if we put `eps` something like 0.1 0.01 0.001 we can bypass spikes
> 
> I'm using a very small `eps`, the default one in llm-foundy yaml files.
> 
> ```
> optimizer:
>   name: decoupled_adamw
>   lr: 2.0e-4
>   betas:
>   - 0.9
>   - 0.95
>   eps: 1.0e-08
>   weight_decay: 0.0
> ```
> 
> Good to know a larger `eps` could help.

same, I will try larger `eps` today

**Comment by hankcs** [comment]
> > > hi @hankcs which Adam params are you using? have you changed `eps`? as I know, if we put `eps` something like 0.1 0.01 0.001 we can bypass spikes
> > 
> > 
> > I'm using a very small `eps`, the default one in llm-foundy yaml files.
> > ```
> > optimizer:
> >   name: decoupled_adamw
> >   lr: 2.0e-4
> >   betas:
> >   - 0.9
> >   - 0.95
> >   eps: 1.0e-08
> >   weight_decay: 0.0
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Good to know a larger `eps` could help.
> 
> same, I will try larger `eps` today

I didn't train so many steps but if I zoom the plot then I can see the loss seems to return to a nomal value after the spike. So maybe it'll be OK after another 1k steps.
<img width="1911" alt="hankcs com 2023-06-12 at 5 02 22 PM" src="https://github.com/mosaicml/llm-foundry/assets/5326890/fb4deae8-a227-436d-aa82-51a701ab3fbd">

**Comment by germanjke** [comment]
> > > > hi @hankcs which Adam params are you using? have you changed `eps`? as I know, if we put `eps` something like 0.1 0.01 0.001 we can bypass spikes
> > > 
> > > 
> > > I'm using a very small `eps`, the default one in llm-foundy yaml files.
> > > ```
> > > optimizer:
> > >   name: decoupled_adamw
> > >   lr: 2.0e-4
> > >   betas:
> > >   - 0.9
> > >   - 0.95
> > >   eps: 1.0e-08
> > >   weight_decay: 0.0
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Good to know a larger `eps` could help.
> > 
> > 
> > same, I will try larger `eps` today
> 
> I didn't train so many steps but if I zoom the plot then I can see the loss seems to return to a nomal value after the spike. So maybe it'll be OK after another 1k steps. <img alt="hankcs com 2023-06-12 at 5 02 22 PM" width="1911" src="https://user-images.githubusercontent.com/5326890/245291124-fb4deae8-a227-436d-aa82-51a701ab3fbd.png">

I have same things with 1b model

<img width="1171" alt="image" src="https://github.com/mosaicml/llm-foundry/assets/43114292/033ce6e2-2622-4901-a020-8e2443284215">

**Comment by germanjke** [comment]
but my 125m models are not ok for sure

<img width="1143" alt="image" src="https://github.com/mosaicml/llm-foundry/assets/43114292/368a3d00-81cc-4b6e-b231-5f36f374ee1f">

**Comment by germanjke** [comment]
@hankcs do you pretrain this things with `attn_impl=triton`? In my experiments only they have same things, and `torch` looks really ok

**Comment by hankcs** [comment]
> 

Yes, you're right! I'm using `attn_impl: triton` for all experiments. `torch` is significantly slower. Not sure if `flash` will be any faster and stabler.

**Comment by germanjke** [comment]
> > 
> 
> Yes, you're right! I'm using `attn_impl: triton` for all experiments. `torch` is significantly slower. Not sure if `flash` will be any faster and stabler.

yes, `torch` is slow, but it's works, at least 0 times see problems with this

so, I think maybe problem in this area, maybe even some hardware missmatch

**Comment by germanjke** [comment]
I have tried both docker images (1.13+11.7 and 2.0+11.8), both have same problems

**Comment by hankcs** [comment]
> > > 
> > 
> > 
> > Yes, you're right! I'm using `attn_impl: triton` for all experiments. `torch` is significantly slower. Not sure if `flash` will be any faster and stabler.
> 
> yes, `torch` is slow, but it's works, at least 0 times see problems with this
> 
> so, I think maybe problem in this area, maybe even some hardware missmatch

Yes, I guess some numerical errors/losses accumulate in `triton`'s implementation.

**Comment by germanjke** [comment]
@vchiley @abhi-mosaic

**Comment by germanjke** [comment]
@vchiley @abhi-mosaic @hanlint

**Comment by germanjke** [comment]
upd: eps increasing helps to stop expode loss, but loss don't converge to before explode level with default eps, so it's not ok solution

**Comment by hankcs** [comment]
> Hi @hankcs,
> 
> It's hard to diagnose convergence issues with certainty, but here are a couple likely causes:
> 
> 1. The lack of weight decay will cause the weights to only ever grow, rather than reach an equilibrium norm. See, e.g., [here](https://dblalock.substack.com/i/126042286/rotational-optimizers-simple-and-robust-dnn-training) for an explanation. You might not see a difference early in training, but I would expect this to cause issues eventually (as you're observing).
> 2. Your other hyperparameters might be causing issues. In particular, momentum [effectively scales the updates by 1/(1 - β2)](https://myrtle.ai/learn/how-to-train-your-resnet-5-hyperparameters/), which is 100 for β2 = .99. For our 1B models, we tend to use β2 = .95, which only scales by 20. Here are some hparams we've used for our MPT 1B runs:
> 
> ```yaml
>   optimizer:
>     name: decoupled_lionw
>     lr: 0.0002
>     betas:
>     - 0.9
>     - 0.95
>     weight_decay: 0.0002
> ```

I didn't get lucky using the hyper params you suggested.

<img width="1925" alt="hankcs com 2023-06-13 at 2 18 30 PM" src="https://github.com/mosaicml/llm-foundry/assets/5326890/50f4662e-7504-4459-9247-adcb91f2ab79">


Maybe there's something not compatible with the A100-PCIE cards I'm using.

**Comment by samhavens** [comment]
If you haven't seen it, @vchiley's [comment on another issue](https://github.com/mosaicml/llm-foundry/issues/284) is relevant here. Batch size, learning rate, data distribution are all relevant, in addition to the attention implementation and optimizer settings.

**Comment by hankcs** [comment]
> If you haven't seen it, @vchiley's [comment on another issue](https://github.com/mosaicml/llm-foundry/issues/284) is relevant here. Batch size, learning rate, data distribution are all relevant, in addition to the attention implementation and optimizer settings.

I don't think the problem of LION is the same with #284 because the LION loss is somewhat flat rather than spike. I'll post results of `attn_impl: torch` here.

**Comment by samhavens** [comment]
This might be a difference of terms, those look like spikes to me, just ones that don't recover at all. But, I thought #284 was relevant because of the comment:

> yes, torch is slow, but it's works, at least 0 times see problems with this

To bring some closure to this issue, we don't think there is a bug in Lion and I don't see evidence here that there is one. It's pretty new, so there isn't a ton of material on using and tuning it, but [lucidrains has some advice on his implementation](https://github.com/lucidrains/lion-pytorch).

I will close this now, with a warning to anyone who finds this that Lion is somewhat sensitive, and to assume that you will be experimenting with schedule and batch size etc if you use it.

**Comment by dblalock** [comment]
Hmm...what dataset, tokenizer, and learning rate schedule are you using? The config I took those hparams from is for training on c4 tokenized with `EleutherAI/gpt-neox-20b` using cosine decay after 5k batches of warmup. I'd be surprised if it's a hardware difference given that hardware runs nearly identically aside from tiny floating point details.

Do you have a link to your full training configuration plus any changes you made to the training script?
